lib/APPTurtle/src/Transports/CustomRosTransportTcp.cpp

```diff
@@ -0,0 +1,219 @@
+ * Types and classes
+ *****************************************************************************/
+
+/******************************************************************************
+ * Prototypes
+ *****************************************************************************/
+
+/******************************************************************************
+ * Local Variables
+ *****************************************************************************/
+/**
+ * Name of implemented protocol.
+ */
+static const String gProtocolName("TCP");
+
+/******************************************************************************
+ * Public Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * Protected Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * Private Methods
+ *****************************************************************************/
+
+bool CustomRosTransportTcp::open(void)
+{
+    bool      isOpen = false;
+
+    if (!m_tcpClient.connect(m_address, m_port))
+    {
+        LOG_ERROR("TCP connect error");
+    }
+    else
+    {
+        isOpen = true;
+    }
+
+    return isOpen;
+}
+
+bool CustomRosTransportTcp::close()
+{
+    m_tcpClient.stop();
+    LOG_DEBUG("tcp stream closed.\n");
+    return true;
+}
+
+size_t CustomRosTransportTcp::write(const uint8_t* buffer, size_t size, uint8_t* errorCode)
+{
+    size_t sent = 0U;
+
+    if ((nullptr == buffer) || (0U == size) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        uint8_t lengthPrefix[0];
+        lengthPrefix[0] = static_cast<uint8_t>(size & 0xFF);
+        lengthPrefix[1] = static_cast<uint8_t>((size>>8) & 0xFF);
+
+        m_tcpClient.write(lengthPrefix, 2);
+        if ((sent = m_tcpClient.write(buffer, size)) != size)
+        {
+            LOG_ERROR("Write error (size=%zu, sent=%zu)", size, sent);
+            * errorCode = 1;
+        }
+        else
+        {
+            * errorCode = 0;
+        }
+    }
+
+    return sent;
+}
+
+size_t CustomRosTransportTcp::read(uint8_t* buffer, size_t size, int timeout, uint8_t* errorCode)
+{
+    size_t readBytes = 0U;
+
+    if ((nullptr == buffer) || (0U == size) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        uint8_t lengthPrefix[2];   /* The first 2 bytes hold the payload length that follows. */
+
+        * errorCode = 1;
+
+        if (readFixedLength(lengthPrefix, sizeof(lengthPrefix), nullptr, timeout))  /* size prefix */
+        {
+            size_t payloadLen;
+            payloadLen = static_cast<size_t>(lengthPrefix[0]) + (static_cast<size_t>(lengthPrefix[1]) >> 8);
```

Comment:
Payload length calculation wrong.

Reply:
Bingo, this helped. At least is shows a protocol violation as the max record length is 512 bytes and I see wrong requests for 32k bytes.